File: bindings/c/tests/e2e.c
/* End-to-end tests for the C client against a spawned bridge (FakeFacade).
 *
 * Set LLMCORE_BRIDGE_PYTHON to a python with llmcore[bridge] importable
 * (default python3). Unix-only (fork/exec/kill). */
#define _POSIX_C_SOURCE 200809L
#include <arpa/inet.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

#include "llmcore_client.h"

static int g_failures = 0;
#define CHECK(cond)                                                  \
  do {                                                               \
    if (!(cond)) {                                                   \
      fprintf(stderr, "FAIL: %s @ line %d\n", #cond, __LINE__);      \
      ++g_failures;                                                  \
    }                                                                \
  } while (0)

static int free_port(void) {
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  struct sockaddr_in a;
  memset(&a, 0, sizeof(a));
  a.sin_family = AF_INET;
  a.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
  a.sin_port = 0;
  bind(fd, (struct sockaddr *)&a, sizeof(a));
  socklen_t L = sizeof(a);
  getsockname(fd, (struct sockaddr *)&a, &L);
  int port = ntohs(a.sin_port);
  close(fd);
  return port;
}

static void msleep(int ms) {
  struct timespec ts;
  ts.tv_sec = ms / 1000;
  ts.tv_nsec = (long)(ms % 1000) * 1000000L;
  nanosleep(&ts, NULL);
}

struct concat {
  char buf[512];
  int done;
};

static int concat_cb(const char *text, int done, void *user) {
  struct concat *a = (struct concat *)user;
  if (done) {
    a->done = 1;
  } else {
    size_t used = strlen(a->buf);
    strncat(a->buf, text, sizeof(a->buf) - used - 1);
  }
  return 0;
}

static int cancel_cb(const char *text, int done, void *user) {
  (void)text;
  (void)done;
  int *n = (int *)user;
  (*n)++;
  return 1; /* cancel on first frame */
}

struct bridge {
  pid_t pid;
  char base[64];
};

static struct bridge start_bridge(int with_audio) {
  struct bridge b;
  b.pid = -1;
  int hport = free_port();
  snprintf(b.base, sizeof(b.base), "http://127.0.0.1:%d", hport);
  char http_addr[64];
  snprintf(http_addr, sizeof(http_addr), "127.0.0.1:%d", hport);

  const char *py = getenv("LLMCORE_BRIDGE_PYTHON");
  if (!py) py = "python3";

  pid_t pid = fork();
  if (pid == 0) {
    setenv("LLMCORE_BRIDGE_FAKE", "1", 1);
    if (with_audio) setenv("LLMCORE_BRIDGE_FAKE_AUDIO", "1", 1);
    char *argv[] = {(char *)py,
                    "-m",
                    "llmcore.bridge.cli",
                    "serve",
                    "--transport",
                    "http",
                    "--http-address",
                    http_addr,
                    "--insecure",
                    "--log-level",
                    "WARNING",
                    NULL};
    execvp(py, argv);
    _exit(127);
  }
  b.pid = pid;

  llmcore_client *c = llmcore_client_new(b.base);
  for (int i = 0; i < 125; i++) { /* up to ~25s */
    int ok = 0;
    llmcore_error *e = llmcore_health(c, &ok);
    if (e) {
      llmcore_error_free(e);
    } else if (ok) {
      llmcore_client_free(c);
      return b;
    }
    msleep(200);
  }
  llmcore_client_free(c);
  fprintf(stderr, "bridge not ready\n");
  kill(pid, SIGTERM);
  exit(2);
}

int main(void) {
  struct bridge b = start_bridge(0);
  llmcore_client *c = llmcore_client_new(b.base);

  /* ensure_compatible accept */
  {
    const char *caps[] = {"tier0"};
    llmcore_error *e = llmcore_ensure_compatible(c, caps, 1);
    CHECK(e == NULL);
    if (e) llmcore_error_free(e);
  }
  /* ensure_compatible reject */
  {
    const char *caps[] = {"tier2.audio"};
    llmcore_error *e = llmcore_ensure_compatible(c, caps, 1);
    CHECK(e != NULL);
    if (e) {
      CHECK(strcmp(e->code, "capability.missing") == 0);
      llmcore_error_free(e);
    }
  }
  /* chat + usage */
  {
    llmcore_chat_result r;
    llmcore_error *e = llmcore_chat(c, "hello world", &r);
    CHECK(e == NULL);
    if (!e) {
      CHECK(strcmp(r.text, "echo: hello world") == 0);
      CHECK(r.prompt_tokens == 2);
      llmcore_chat_result_free(&r);
    } else {
      llmcore_error_free(e);
    }
  }
  /* stream concatenation */
  {
    struct concat a;
    memset(&a, 0, sizeof(a));
    llmcore_error *e = llmcore_chat_stream(c, "stream this please", concat_cb, &a);
    CHECK(e == NULL);
    if (e) llmcore_error_free(e);
    CHECK(a.done == 1);
    CHECK(strcmp(a.buf, "echo: stream this please") == 0);
  }
  /* count */
  {
    int n = 0;
    llmcore_error *e = llmcore_count_tokens(c, "one two three four", &n);
    CHECK(e == NULL);
    if (e) llmcore_error_free(e);
    CHECK(n == 4);
  }
  /* cost */
  {
    double tc = 0.0;
    char *cur = NULL;
    llmcore_error *e = llmcore_estimate_cost(c, "fake", "fake-1", 1000000, 1000000, &tc, &cur);
    CHECK(e == NULL);
    if (e) llmcore_error_free(e);
    CHECK(cur && strcmp(cur, "USD") == 0);
    CHECK(tc == 3.0);
    free(cur);
  }
  /* catalog: providers */
  {
    char **it = NULL;
    size_t n = 0;
    llmcore_error *e = llmcore_list_providers(c, &it, &n);
    CHECK(e == NULL);
    if (e) llmcore_error_free(e);
    CHECK(n == 1 && it && strcmp(it[0], "fake") == 0);
    llmcore_string_array_free(it, n);
  }
  /* catalog: models */
  {
    char **it = NULL;
    size_t n = 0;
    llmcore_error *e = llmcore_list_models(c, "fake", &it, &n);
    CHECK(e == NULL);
    if (e) llmcore_error_free(e);
    CHECK(n == 2 && strcmp(it[0], "fake-1") == 0 && strcmp(it[1], "fake-2") == 0);
    llmcore_string_array_free(it, n);
  }
  /* embed unsupported */
  {
    const char *in[] = {"x"};
    llmcore_error *e = llmcore_embed(c, in, 1);
    CHECK(e != NULL);
    if (e) {
      CHECK(strcmp(e->category, "ERROR_CATEGORY_UNSUPPORTED") == 0);
      llmcore_error_free(e);
    }
  }
  /* provider rate-limit structured decode */
  {
    llmcore_chat_result r;
    llmcore_error *e = llmcore_chat(c, "__error__:provider_rate_limited", &r);
    CHECK(e != NULL);
    if (e) {
      CHECK(strcmp(e->category, "ERROR_CATEGORY_PROVIDER") == 0);
      CHECK(strcmp(e->code, "provider.rate_limited") == 0);
      CHECK(e->http_status == 429);
      CHECK(e->retryable == 1);
      CHECK(e->retry_after_ms == 2000.0);
      CHECK(e->provider && strcmp(e->provider, "fake") == 0);
      llmcore_error_free(e);
    } else {
      llmcore_chat_result_free(&r);
    }
  }
  /* stream cancellation */
  {
    int cnt = 0;
    llmcore_error *e = llmcore_chat_stream(c, "__cancel__", cancel_cb, &cnt);
    CHECK(e == NULL); /* user cancel reported as success */
    if (e) llmcore_error_free(e);
    CHECK(cnt >= 1);
  }

  /* ---- audio (Tier 2): a second, audio-enabled bridge ---- */
  {
    struct bridge ab = start_bridge(1);
    llmcore_client *ac = llmcore_client_new(ab.base);

    /* synthesize: audio_data decodes to "tts:hello" */
    {
      llmcore_speech_result r;
      llmcore_error *e = llmcore_synthesize(ac, "hello", &r);
      CHECK(e == NULL);
      if (e) {
        llmcore_error_free(e);
      } else {
        CHECK(r.audio_len == 9 && memcmp(r.audio, "tts:hello", 9) == 0);
        CHECK(strcmp(r.model, "fake-tts") == 0);
        llmcore_speech_result_free(&r);
      }
    }
    /* transcribe: audio bytes round-trip to text */
    {
      llmcore_transcription_result r;
      llmcore_error *e =
          llmcore_transcribe(ac, (const unsigned char *)"hello world", 11, &r);
      CHECK(e == NULL);
      if (e) {
        llmcore_error_free(e);
      } else {
        CHECK(strcmp(r.text, "hello world") == 0);
        CHECK(strcmp(r.language, "en") == 0);
        llmcore_transcription_result_free(&r);
      }
    }
    /* generate_image: n=2, data[0] == base64("img:a cat") */
    {
      llmcore_image_result r;
      llmcore_error *e = llmcore_generate_image(ac, "a cat", 2, &r);
      CHECK(e == NULL);
      if (e) {
        llmcore_error_free(e);
      } else {
        CHECK(r.n_images == 2);
        if (r.n_images >= 1) CHECK(strcmp(r.images[0], "aW1nOmEgY2F0") == 0);
        CHECK(strcmp(r.model, "fake-img") == 0);
        llmcore_image_result_free(&r);
      }
    }
    /* ocr: bytes -> model/pages_processed/doc_size_bytes */
    {
      llmcore_ocr_result r;
      llmcore_error *e =
          llmcore_ocr(ac, (const unsigned char *)"PDFBYTES", 8, NULL, &r);
      CHECK(e == NULL);
      if (e) {
        llmcore_error_free(e);
      } else {
        CHECK(strcmp(r.model, "fake-ocr") == 0);
        CHECK(r.pages_processed == 1);
        CHECK(r.has_doc_size_bytes && r.doc_size_bytes == 8);
        llmcore_ocr_result_free(&r);
      }
    }
    /* analyze_text (no features): model set, no summary */
    {
      llmcore_text_analysis_result r;
      llmcore_error *e = llmcore_analyze_text(ac, "some text", &r);
      CHECK(e == NULL);
      if (e) {
        llmcore_error_free(e);
      } else {
        CHECK(r.has_summary == 0);
        CHECK(strcmp(r.model, "fake-analyze") == 0);
        llmcore_text_analysis_result_free(&r);
      }
    }

    llmcore_client_free(ac);
    kill(ab.pid, SIGTERM);
    int ast;
    waitpid(ab.pid, &ast, 0);
  }

  llmcore_client_free(c);
  kill(b.pid, SIGTERM);
  int st;
  waitpid(b.pid, &st, 0);

  if (g_failures == 0) {
    printf("ALL TESTS PASSED\n");
    return 0;
  }
  printf("%d FAILURES\n", g_failures);
  return 1;
}

File: bindings/rust/llmcore-client/tests/e2e.rs
//! End-to-end tests against a spawned `llmcore-bridge` (FakeFacade).
//!
//! Set `LLMCORE_BRIDGE_PYTHON` to a python with `llmcore[bridge]` importable
//! (default `python3`). Unix-only harness (kills via the OS).

use std::net::TcpListener;
use std::process::{Child, Command};
use std::time::{Duration, Instant};

use llmcore_client::{v1, LlmcoreClient};

fn free_port() -> u16 {
    TcpListener::bind("127.0.0.1:0")
        .unwrap()
        .local_addr()
        .unwrap()
        .port()
}

/// A spawned bridge process; killed on drop.
struct Bridge {
    child: Child,
    endpoint: String,
}

impl Bridge {
    async fn start() -> Bridge {
        Self::start_env(&[]).await
    }

    /// Start with the Tier-2 fake audio surface enabled (advertises tier2.audio).
    async fn start_audio() -> Bridge {
        Self::start_env(&[("LLMCORE_BRIDGE_FAKE_AUDIO", "1")]).await
    }

    async fn start_env(extra_env: &[(&str, &str)]) -> Bridge {
        let grpc_port = free_port();
        let http_port = free_port();
        let grpc_addr = format!("127.0.0.1:{}", grpc_port);
        let http_addr = format!("127.0.0.1:{}", http_port);
        let py = std::env::var("LLMCORE_BRIDGE_PYTHON").unwrap_or_else(|_| "python3".to_string());

        let mut cmd = Command::new(py);
        cmd.args([
            "-m",
            "llmcore.bridge.cli",
            "serve",
            "--transport",
            "grpc,http",
            "--grpc-address",
            &grpc_addr,
            "--http-address",
            &http_addr,
            "--insecure",
            "--log-level",
            "WARNING",
        ])
        .env("LLMCORE_BRIDGE_FAKE", "1");
        for (k, v) in extra_env {
            cmd.env(k, v);
        }
        let child = cmd.spawn().expect("spawn bridge");

        let endpoint = format!("http://{}", grpc_addr);
        let deadline = Instant::now() + Duration::from_secs(25);
        loop {
            assert!(Instant::now() < deadline, "bridge not ready");
            if let Ok(mut c) = LlmcoreClient::connect(endpoint.clone()).await {
                if let Ok(h) = c.health().await {
                    if h.ok {
                        break;
                    }
                }
            }
            tokio::time::sleep(Duration::from_millis(200)).await;
        }
        Bridge { child, endpoint }
    }
}

impl Drop for Bridge {
    fn drop(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

async fn connected() -> (LlmcoreClient, Bridge) {
    let b = Bridge::start().await;
    let c = LlmcoreClient::connect(b.endpoint.clone()).await.unwrap();
    (c, b)
}

#[tokio::test]
async fn ensure_compatible() {
    let (mut c, _b) = connected().await;
    let info = c.ensure_compatible(&["tier0"]).await.unwrap();
    assert_eq!(info.contract_version, "llmcore.v1");

    let err = c.ensure_compatible(&["tier2.audio"]).await.unwrap_err();
    assert_eq!(err.code, "capability.missing");
}

#[tokio::test]
async fn chat_and_usage() {
    let (mut c, _b) = connected().await;
    let res = c
        .chat(v1::ChatRequest {
            message: "hello world".into(),
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(res.text, "echo: hello world");
    assert_eq!(res.usage.unwrap().prompt_tokens, Some(2));
}

#[tokio::test]
async fn chat_stream_concat() {
    let (mut c, _b) = connected().await;
    let mut s = c
        .chat_stream(v1::ChatRequest {
            message: "stream this please".into(),
            ..Default::default()
        })
        .await
        .unwrap();
    let mut parts = String::new();
    let mut done = false;
    while let Some(chunk) = s.message().await.unwrap() {
        if chunk.done {
            done = true;
        } else {
            parts.push_str(&chunk.text);
        }
    }
    assert!(done);
    assert_eq!(parts, "echo: stream this please");
}

#[tokio::test]
async fn count_and_cost() {
    let (mut c, _b) = connected().await;
    let ct = c
        .count_tokens(v1::CountTokensRequest {
            text: "one two three four".into(),
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(ct.tokens, 4);

    let ce = c
        .estimate_cost(v1::EstimateCostRequest {
            provider_name: "fake".into(),
            model_name: "fake-1".into(),
            prompt_tokens: 1_000_000,
            completion_tokens: 1_000_000,
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(ce.currency, "USD");
    assert_eq!(ce.total_cost, 3.0);
}

#[tokio::test]
async fn catalog() {
    let (mut c, _b) = connected().await;
    assert_eq!(c.list_providers().await.unwrap().providers, vec!["fake"]);
    assert_eq!(
        c.list_models("fake").await.unwrap().models,
        vec!["fake-1", "fake-2"]
    );
    let d = c.get_provider_details(Some("fake".into())).await.unwrap();
    assert_eq!(d.id, "fake-1");
    assert_eq!(d.context_length, 8192);
}

#[tokio::test]
async fn embed_unsupported() {
    let (mut c, _b) = connected().await;
    let err = c
        .embed(v1::EmbedRequest {
            input: vec!["x".into()],
            ..Default::default()
        })
        .await
        .unwrap_err();
    assert_eq!(err.category, "ERROR_CATEGORY_UNSUPPORTED");
}

#[tokio::test]
async fn provider_rate_limit_decode() {
    let (mut c, _b) = connected().await;
    let err = c
        .chat(v1::ChatRequest {
            message: "__error__:provider_rate_limited".into(),
            ..Default::default()
        })
        .await
        .unwrap_err();
    assert_eq!(err.category, "ERROR_CATEGORY_PROVIDER");
    assert_eq!(err.code, "provider.rate_limited");
    assert_eq!(err.http_status, Some(429));
    assert!(err.retryable);
    assert_eq!(err.retry_after_ms, Some(2000.0));
    assert_eq!(err.provider.as_deref(), Some("fake"));
}

#[tokio::test]
async fn stream_cancel() {
    let (mut c, _b) = connected().await;
    let mut s = c
        .chat_stream(v1::ChatRequest {
            message: "__cancel__".into(),
            ..Default::default()
        })
        .await
        .unwrap();
    let first = s.message().await.unwrap();
    assert!(first.is_some());
    s.cancel();
    assert!(s.message().await.unwrap().is_none());
}

// ---- audio (Tier 2) ---------------------------------------------------- //
//
// Enum fields are compared by their verified numeric proto values (noted inline)
// to stay robust to prost's enum-variant naming. Oneof frames use the generated
// oneof modules (`audio_in::Frame`, `synth_control::Frame`,
// `voice_agent_client_event::Event`, `ocr_request::Source`) with single-word
// variant names derived from the oneof field names.

async fn connected_audio() -> (LlmcoreClient, Bridge) {
    let b = Bridge::start_audio().await;
    let c = LlmcoreClient::connect(b.endpoint.clone()).await.unwrap();
    (c, b)
}

#[tokio::test]
async fn audio_capabilities_advertised() {
    let (mut c, _b) = connected_audio().await;
    let info = c.get_info().await.unwrap();
    for cap in [
        "tier2.audio",
        "audio.transcribe_stream",
        "audio.synthesize_stream",
        "audio.voice_agent",
        "audio.synthesize",
        "audio.transcribe",
        "audio.generate_image",
        "audio.ocr",
        "audio.analyze_text",
    ] {
        assert!(
            info.capabilities.iter().any(|x| x == cap),
            "missing capability {cap}"
        );
    }
}

#[tokio::test]
async fn transcribe_stream() {
    let (mut c, _b) = connected_audio().await;
    let reqs = tokio_stream::iter(vec![
        v1::AudioIn {
            frame: Some(v1::audio_in::Frame::Audio(b"hello".to_vec())),
        },
        v1::AudioIn {
            frame: Some(v1::audio_in::Frame::Audio(b"world".to_vec())),
        },
        v1::AudioIn {
            frame: Some(v1::audio_in::Frame::Control(3)), // STT_CONTROL_CLOSE
        },
    ]);
    let mut s = c.transcribe_stream(reqs).await.unwrap();
    let mut types: Vec<i32> = Vec::new();
    let mut final_text = String::new();
    while let Some(ev) = s.message().await.unwrap() {
        types.push(ev.r#type);
        if ev.r#type == 2 {
            // FINAL
            final_text = ev.text.clone();
        }
    }
    // INTERIM, INTERIM, FINAL, UTTERANCE_END
    assert_eq!(types, vec![1, 1, 2, 3]);
    assert_eq!(final_text, "hello world");
}

#[tokio::test]
async fn synthesize_stream() {
    let (mut c, _b) = connected_audio().await;
    let pieces = ["foo", "bar", "baz"];
    let mut frames: Vec<v1::SynthControl> = pieces
        .iter()
        .map(|p| v1::SynthControl {
            frame: Some(v1::synth_control::Frame::Text((*p).to_string())),
        })
        .collect();
    frames.push(v1::SynthControl {
        frame: Some(v1::synth_control::Frame::Control(3)), // TTS_CONTROL_CLOSE
    });
    let mut s = c.synthesize_stream(tokio_stream::iter(frames)).await.unwrap();
    let mut chunks: Vec<String> = Vec::new();
    let mut seqs: Vec<i64> = Vec::new();
    while let Some(out) = s.message().await.unwrap() {
        chunks.push(String::from_utf8(out.audio).unwrap());
        seqs.push(out.seq);
    }
    assert_eq!(chunks, vec!["foo", "bar", "baz"]);
    assert_eq!(seqs, vec![0, 1, 2]);
}

#[tokio::test]
async fn voice_agent_duplex() {
    let (mut c, _b) = connected_audio().await;
    // A non-settings leading frame -> the default provider (the fake).
    let reqs = tokio_stream::iter(vec![
        v1::VoiceAgentClientEvent {
            event: Some(v1::voice_agent_client_event::Event::InjectUserMessage(
                "hi there".into(),
            )),
        },
        v1::VoiceAgentClientEvent {
            event: Some(v1::voice_agent_client_event::Event::Audio(vec![1, 2])),
        },
    ]);
    let mut s = c.voice_agent(reqs).await.unwrap();
    let mut events = Vec::new();
    while let Some(ev) = s.message().await.unwrap() {
        events.push(ev);
    }
    assert!(!events.is_empty());
    assert_eq!(events.first().unwrap().r#type, 1); // WELCOME
    assert_eq!(events.last().unwrap().r#type, 17); // CLOSE
    let conv = events
        .iter()
        .find(|e| e.r#type == 3) // CONVERSATION_TEXT
        .expect("conversation-text event");
    assert_eq!(conv.role.as_deref(), Some("user"));
    assert_eq!(conv.content.as_deref(), Some("hi there"));
    let audio = events
        .iter()
        .find(|e| e.r#type == 8) // AUDIO
        .expect("audio event");
    assert_eq!(audio.audio.as_deref(), Some(&b"agent:\x01\x02"[..]));
}

#[tokio::test]
async fn synthesize_unary() {
    let (mut c, _b) = connected_audio().await;
    let r = c
        .synthesize(v1::SynthesizeRequest {
            text: "hello".into(),
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(r.audio_data, b"tts:hello");
    assert_eq!(r.model, "fake-tts");
}

#[tokio::test]
async fn transcribe_unary() {
    let (mut c, _b) = connected_audio().await;
    let r = c
        .transcribe(v1::TranscribeRequest {
            audio_data: b"hello world".to_vec(),
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(r.text, "hello world");
    assert_eq!(r.language.as_deref(), Some("en"));
    assert_eq!(r.segments.len(), 1);
    assert_eq!(r.segments[0].speaker.as_deref(), Some("spk_0"));
}

#[tokio::test]
async fn generate_image_unary() {
    let (mut c, _b) = connected_audio().await;
    let r = c
        .generate_image(v1::GenerateImageRequest {
            prompt: "a cat".into(),
            n: 2,
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(r.images.len(), 2);
    // base64("img:a cat") == "aW1nOmEgY2F0"
    assert_eq!(r.images[0].data.as_deref(), Some("aW1nOmEgY2F0"));
}

#[tokio::test]
async fn ocr_unary() {
    let (mut c, _b) = connected_audio().await;
    let r = c
        .ocr(v1::OcrRequest {
            source: Some(v1::ocr_request::Source::Data(b"PDFBYTES".to_vec())),
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(r.model, "fake-ocr");
    assert_eq!(r.pages_processed, 1);
    assert_eq!(r.doc_size_bytes, Some(8));
    assert_eq!(r.pages.len(), 1);
}

#[tokio::test]
async fn analyze_text_unary() {
    let (mut c, _b) = connected_audio().await;
    // No features -> the fake returns model "fake-analyze", no summary/topics.
    let r = c
        .analyze_text(v1::AnalyzeTextRequest {
            text: "some text".into(),
            ..Default::default()
        })
        .await
        .unwrap();
    assert_eq!(r.model.as_deref(), Some("fake-analyze"));
    assert!(r.summary.is_none());
    assert!(r.topics.is_empty());
}
